providers/gitlab.go

```diff
@@ -206,15 +228,67 @@ func (p *GitLabProvider) EnrichSessionState(ctx context.Context, s *sessions.Ses
 		return fmt.Errorf("user email is not verified")
 	}
 
-	// Check group membership
-	// TODO (@NickMeves) - Refactor to Authorize
-	err = p.verifyGroupMembership(userInfo)
-	if err != nil {
-		return fmt.Errorf("group membership check failed: %v", err)
-	}
+	p.addGroupMembershipToState(ctx, s)
+
+	p.addProjectMembershipToState(ctx, s)
 
 	s.User = userInfo.Username
 	s.Email = userInfo.Email
 
 	return nil
+
+}
+
+// addProjectMembership adds projects into session.Groups
+func (p *GitLabProvider) addGroupMembershipToState(ctx context.Context, s *sessions.SessionState) {
+	if len(p.Groups) == 0 {
+		return
+	}
+
+	// Iterate over projects, check if oauth2-proxy can get project information on behalf of the user
+	for _, group := range p.Groups {
+		s.Groups = append(s.Groups, fmt.Sprintf("group:%s", group))
+	}
+
+}
+
+// addProjectMembership adds projects into session.Groups
+func (p *GitLabProvider) addProjectMembershipToState(ctx context.Context, s *sessions.SessionState) {
+	if len(p.Projects) == 0 {
+		return
+	}
+
+	// Iterate over projects, check if oauth2-proxy can get project information on behalf of the user
+	for _, project := range p.Projects {
+		project, err := p.getProjectInfo(ctx, s, project)
+		// If no error, it's a member of the project
+		if err == nil && !project.Archived {
+			s.Groups = append(s.Groups, fmt.Sprintf("project:%s", project.PathWithNamespace))
+		}
+
+		if err != nil {
+			logger.Printf("Warning: project info request failed: %v", err)
+		}
+
+		if project != nil && project.Archived {
+			logger.Printf("Warning: project %s is archived", project.Name)
+		}
+
+	}
+
+}
+
+// PrefixAllowedGroups return a list of allowed groups, prefixed by their `kind` value
+func (p *GitLabProvider) PrefixAllowedGroups() (groups []string) {
```

Comment:
Just to confirm you are cool with this AuthZ logic:

This will result in ANY style logic between groups & projects together. So if you said `-groups=foobar` and `-projects=baz-project` either would suffice for access rather needing to be in group `foobar` or `baz-project` access.

Does that fit the AuthZ paradigm you envisioned for group vs project access if both are toggled? If so, I'll trust your judgment.

Reply:
Yes, I confirm this is the AuthZ logic I imagine. For feedbacks and other opinions I asked other Gitlab users, they are all agree on the **OR** logic, since if you are a group member, you will probably be a project member too.

Reply:
Nice! I appreciate the community research on this - Thanks!